In [1]:
import pandas as p
import numpy as n

df=p.read_csv("adult.csv")
print("First five rows:", df.head(5))
print("Info:", df.info())

print("Initial missing vals\n", df.isnull().sum())
df.replace("?", n.nan, inplace=True)
print("Missing vals\n", df.isnull().sum())

First five rows:    age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              

In [49]:
missing=df.isnull().sum()
percentage=(missing/len(df))*100

table=p.DataFrame({"Missing count":missing, "Missing percentage":percentage})

table=table.sort_values(by="Missing count",ascending=False)

print(table)

                Missing count  Missing percentage
occupation               1843            5.660146
workclass                1836            5.638647
native.country            583            1.790486
fnlwgt                      0            0.000000
education                   0            0.000000
education.num               0            0.000000
age                         0            0.000000
marital.status              0            0.000000
relationship                0            0.000000
sex                         0            0.000000
race                        0            0.000000
capital.gain                0            0.000000
capital.loss                0            0.000000
hours.per.week              0            0.000000
income                      0            0.000000


In [50]:
#workclass has 1800 missing values and dropping might remove useful data. so impute rather than replace with most common workclass
#occupation also has 1800 and impute for the same reason

df["workclass"] = df["workclass"].fillna("Unknown")
df["occupation"] = df["occupation"].fillna("Unknown")

df

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,Unknown,77053,HS-grad,9,Widowed,Unknown,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,Unknown,186061,Some-college,10,Widowed,Unknown,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,22,Private,310152,Some-college,10,Never-married,Protective-serv,Not-in-family,White,Male,0,0,40,United-States,<=50K
32557,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32558,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32559,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K


In [51]:
print("Duplicate rows\n", df.duplicated().sum())

print("Duplicates after ignoring target\n", df.drop(columns=["income"]).duplicated().sum())

#that one record is not duplicate since income can differ while having same details

Duplicate rows
 24
Duplicates after ignoring target
 25


In [52]:
print("Education\n", df["education"].unique())
print("Marital Status\n", df["marital.status"].unique())
print("Native country\n", df["native.country"].unique())

df["education"] = df["education"].str.strip()
df["marital.status"] = df["marital.status"].str.strip()
df["native.country"] = df["native.country"].str.strip()


print("Education\n", df["education"].unique())
print("Marital Status\n", df["marital.status"].unique())
print("Native country\n", df["native.country"].unique())

Education
 <StringArray>
[     'HS-grad', 'Some-college',      '7th-8th',         '10th',
    'Doctorate',  'Prof-school',    'Bachelors',      'Masters',
         '11th',   'Assoc-acdm',    'Assoc-voc',      '1st-4th',
      '5th-6th',         '12th',          '9th',    'Preschool']
Length: 16, dtype: str
Marital Status
 <StringArray>
[              'Widowed',              'Divorced',             'Separated',
         'Never-married',    'Married-civ-spouse', 'Married-spouse-absent',
     'Married-AF-spouse']
Length: 7, dtype: str
Native country
 <StringArray>
[             'United-States',                          nan,
                     'Mexico',                     'Greece',
                    'Vietnam',                      'China',
                     'Taiwan',                      'India',
                'Philippines',            'Trinadad&Tobago',
                     'Canada',                      'South',
         'Holand-Netherlands',                'Puerto-Rico',
     

In [53]:
print(df.describe())

print(df.describe(include='object'))

                age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  
       workclass education      marital.status      occupation 

C:\Users\Pc\AppData\Local\Temp\ipykernel_14568\1983982935.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print(df.describe(include='object'))


In [54]:
print("occupation count", df["occupation"].value_counts())
print("occupation:", df["occupation"].value_counts().idxmax())
print("sex percent distribution:", df["sex"].value_counts(normalize=True)*100)
print("income:", df["income"].value_counts(normalize=True)*100)

occupation count occupation
Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Unknown              1843
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Name: count, dtype: int64
occupation: Prof-specialty
sex percent distribution: sex
Male      66.920549
Female    33.079451
Name: proportion, dtype: float64
income: income
<=50K    75.919044
>50K     24.080956
Name: proportion, dtype: float64


In [55]:
print(df.groupby("education")["education.num"].unique())

education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object


In [56]:
import sqlite3

df_og=p.read_csv("adult.csv")
conn=sqlite3.connect("adult.db")

df_og.to_sql("adult_income", conn, if_exists="replace", index=False)

query="""select * from adult_income where age>30"""
df_sql=p.read_sql_query(query, conn)

print(df_sql.head())

   age workclass  fnlwgt     education  education.num marital.status  \
0   90         ?   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66         ?  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                  ?  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                  ?      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States

In [57]:
print("Original CSV shape:", df_og.shape)
print("SQL result shape:", df_sql.shape)

Original CSV shape: (32561, 15)
SQL result shape: (21989, 15)


In [58]:
print("Minimum age in SQL result:", df_sql["age"].min())

Minimum age in SQL result: 31


In [59]:
df_check = df_og[df_og["age"] > 30]

print("Pandas filtered shape:", df_check.shape)
print("SQL result shape:", df_sql.shape)

Pandas filtered shape: (21989, 15)
SQL result shape: (21989, 15)


# Data Profiling Summary Report

## 1. Dataset Overview

The Adult Income dataset contains 32,561 rows and 15 columns. It represents demographic, education, employment, and financial information about individuals, with the target variable `income` indicating whether a person's annual income is `<=50K` or `>50K`.

## 2. Data Quality Issues Found

- Missing values: Missing values were initially represented by `"?"`, so they were replaced with `NaN`.
- Workclass: 1,836 missing values (5.64%).
- Occupation: 1,843 missing values (5.66%).
- Native-country: 583 missing values (1.79%).
- All other columns had 0 missing values.
- Duplicates: Exact duplicate rows were checked using `duplicated().sum()`. Duplicates ignoring the target column were also checked. Records with identical features but different income labels were not treated as duplicates because they can represent valid different individuals.
- Inconsistent categorical values: Leading/trailing whitespace was identified as a potential inconsistency in categorical columns and was removed using `.str.strip()`.
- Education consistency: `education` and `education.num` were checked using `groupby()` and `.unique()`. Each education level mapped to a single education number, so no inconsistencies were found.

## 3. Key Observations

- The mean age is 38.58 years, while the median is 37 years, indicating a slight right skew.
- `hours.per.week` has a mean of 40.44 hours and a median of 40 hours, so there is very little difference between them.
- Prof-specialty is the most common occupation, with 4,140 records.
- The sex distribution is 66.92% Male and 33.08% Female.
- The income classes are imbalanced:
  - `<=50K`: 75.92%
  - `>50K`: 24.08%
- Therefore, the `<=50K` class is the clear majority class and class imbalance should be considered during machine-learning model evaluation.

## 4. Cleaning Decisions Made

- Replaced all `"?"` values with proper `NaN` values so that missing data could be detected correctly.
- Missing `workclass` and `occupation` values were imputed with `"Unknown"` rather than dropping the rows, because these columns contain about 5–6% missing values and the remaining information in those records is still useful.
- No numeric columns required imputation because there were no missing numeric values.
- Leading/trailing whitespace was removed from categorical columns using `.str.strip()` to ensure consistent category names.
- Exact duplicate rows were considered duplicates because all their attributes and target values were identical. Rows that were identical except for `income` were retained because different income outcomes can be valid observations.